include/mfmg/concepts.hpp

```diff
@@ -14,29 +14,35 @@
 
 #include <mfmg/exceptions.hpp>
 
+#include <boost/property_tree/ptree.hpp>
+
 #include <memory>
+#include <mpi.h>
 
 namespace mfmg
 {
 
-template <class VectorType>
+template <typename VectorType>
 class Operator
 {
 public:
   using operator_type = Operator<VectorType>;
   using vector_type = VectorType;
 
 public:
-  void apply(const vector_type &x, vector_type &y) const;
-  std::shared_ptr<operator_type> transpose() const;
-  std::shared_ptr<operator_type> multiply(const operator_type &b) const;
+  virtual size_t m() const = 0;
+  virtual size_t n() const = 0;
+  virtual void apply(vector_type const &x, vector_type &y) const = 0;
+  virtual std::shared_ptr<operator_type> transpose() const = 0;
```

Comment:
The concept of operator should really only have the meaning of how to produce right hand side by operating on left hand side. Its interface should clearly indicate that. Thus, I believe that member functions like `transpose()` and `multiply()` do not belong in it, as they somewhat restrict the concept to operators corresponding to matrices. It also becomes a rabbit hole if we start creating operators with more required functions and putting them in `Operator`. Therefore, I'd like to retain the old interface with `apply` and `build_{range,domain}_vector`. We could even take `m()` and `n()` out but that would most likely require `Vector` concept.

Reply:
What I would propose to do instead is to create a concept of `MatrixOperator` which inherits `Operator`. Then, in `DealIIMatrixOperator` and `TrilinosMatrixOperator` could inherit from that instead.

Reply:
Yeah, good idea.